### PR TITLE
Introduce a new release process

### PR DESCRIPTION
**Pull request by kylekthompson**
Introduces a new release process for Captain. Instead of tracking release branches (for production), we'll **trigger a workflow** which will:

- Create a draft release (or overwrite an existing draft release, if we need to re-cut a release)
- Build the binaries
- Upload those binaries to the draft release
- Wait for approval 
  - For SOC2 and to give an opportunity to tweak the release notes
- Publish that release
- Update (or create) an aliased major-version release
- Upload the binaries from the new version's release to the major-version release
- Upload the binaries from the new version's release to cloudfront (S3)

Additionally, this workflow supports `unstable` and `testing` releases. This PR doesn't introduce it yet, but on merges to main, we'll trigger this workflow to create an `unstable` release tagged with the commit sha and an aliased release `unstable` which always points to the most recent commit on main. The `testing` release allows for engineers to create `testing-SHA` releases based on commits in their branches if necessary. `unstable` and `testing` releases do not require approval.

Noteworthy: we no longer upload `versions.json`. I checked datadog and no one is using this now.

I've tested this fairly extensively on my fork- the only bits that _haven't_ been actually invoked are the `aws s3 cp` and `aws cloudfront create-invalidation` calls (since I didn't bother setting up a place to test that; I did, however, change them to echos to check the arguments).

### Waiting for approval
Screenshot 2023-04-11 at 11 16 26 AM

### Approved `production` release
Screenshot 2023-04-11 at 1 56 31 PM

### Rejected `production` release
Screenshot 2023-04-11 at 1 56 42 PM

### `unstable` release
Screenshot 2023-04-11 at 1 57 02 PM

### `testing` release
Screenshot 2023-04-11 at 1 58 44 PM

### CODE REVIEW

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:32]: is this possible? dead code?

File: .github/workflows/continuous_integration.yaml
Comment (michaelglass) [inline .github/workflows/continuous_integration.yaml:18]: 💭  we use `git describe` (after a first tag is set) to make these versions more readable

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:87]: You don't intend to push all tags do you? If not, you just need to push the tag name. I guess it's mostly safe but there's some time-of-check time-of-use issue if two folks are releasing at similar times.
```suggestion
          git push --force origin ${{ needs.extract-version-details.outputs.full-version }}
```

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:99]: 💭 I don't think this is doing anything. If the tag moves the release should move with it (yesterday when I force pushed the v1 tag, the v1 release changed automatically)

File: .github/workflows/release.yaml
Comment (michaelglass) [reply .github/workflows/release.yaml:89]: (the gh release create command will happily create multiple releases with the same inputs for me)

Comment (michaelglass) [reply .github/workflows/release.yaml:89]: it does fail when creating multiple non-draft releases on the same tag

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:130]: 💭 maybe worth merging these if the only difference is the binary name

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:164]: that's a neat trick. I didn't realize you could do that.

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:86]: 💭 if we use `git tag -a -m""` or `git tag -s -m""` (if that works out of the box? signing is annoying) then we could use `git describe`

File: .github/workflows/buildkite_agent.yaml
Comment (michaelglass) [inline .github/workflows/buildkite_agent.yaml:19]: should we upstream these settings into the magefile?

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:189]: might warrant a comment here about this being empty for testing releases or asserting on `inputs.kind` to clarify why this conditional is there

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:197]: ~again, I don't think this does anything

File: .github/workflows/release.yaml
Comment (michaelglass) [reply .github/workflows/release.yaml:164]: (presumably at this stage you edit the release notes? before approval)

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:213]: 💭 I don't understand this conditional. The always() feels at odds with the success requirements

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:256]: 💭 might be worth adding a comment indicating the extension is intentionally missing

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:262]: 😱  `}}}` (looks right though)

Comment (michaelglass) [reply .github/workflows/release.yaml:262]: 💭 I bet this little script would look a little nicer if you set the github vars to env vars

File: magefile.go
Comment (michaelglass) [inline magefile.go:215]: 💭 considering we're already parsing and re-routing these vars, perhaps it's more intuitive / ergonomic to give ourself a special env var for setting captain version, like `$CAPTAIN_VERSION` ?

**Review by michaelglass — DISMISSED** [review]
looks great!

handful of suggestions but I don't think anything is blocking

File: .github/workflows/continuous_integration.yaml
Comment (kylekthompson) [reply .github/workflows/continuous_integration.yaml:18]: We don't always have a tag here- this is just the locally built version of the CLI, but Captain Cloud requires the versions set in the user agent to adhere to a certain format (or I wouldn't even bother setting it)

File: .github/workflows/release.yaml
Comment (kylekthompson) [reply .github/workflows/release.yaml:87]: Honestly I grabbed this a little blindly from the old release. This sounds like a good change to make now that we have a few kinds of releases (even if those are, within their kind, running in concurrency controls)

File: .github/workflows/continuous_integration.yaml
Comment (michaelglass) [reply .github/workflows/continuous_integration.yaml:18]: `git describe` work without a tag set (as long as there's a tag in the git history)

File: .github/workflows/release.yaml
Comment (kylekthompson) [reply .github/workflows/release.yaml:89]: > the gh release create command will happily create multiple draft releases with the same inputs for me

Oh interesting- I must've been accidentally testing w/ a published release 🤔 I'll poke more at the `gh` CLI locally to confirm the behavior we want

Comment (kylekthompson) [reply .github/workflows/release.yaml:89]: Specifically, I was trying to get re-cutting releases to work in the event of an approval rejection/some issue building

File: .github/workflows/release.yaml
Comment (kylekthompson) [reply .github/workflows/release.yaml:130]: This was yanked from the old script. I'll take a look

File: .github/workflows/release.yaml
Comment (kylekthompson) [reply .github/workflows/release.yaml:164]: Yep! (or after, I guess, but probably at this point)

File: .github/workflows/release.yaml
Comment (kylekthompson) [reply .github/workflows/release.yaml:189]: Is it not clear as-is? 🤔 If there is no aliased-version set, don't push aliased tags and make an aliased release?

I didn't tie this to the kind because it's not so much related to the kind as it is the aliased version (e.g. it doesn't care that for production, it's v1 or unstable it's unstable)

Comment (michaelglass) [reply .github/workflows/release.yaml:189]: when I was reading I asked, "when is the there no aliased tag", "is this defensive code?". until I looked back and realized the it's empty for `testing` releases.

(none of the feedback in this PR should block merge though!)

File: .github/workflows/release.yaml
Comment (kylekthompson) [reply .github/workflows/release.yaml:213]: I can probably leave a comment

File: .github/workflows/release.yaml
Comment (michaelglass) [reply .github/workflows/release.yaml:89]: yeah makes sense

File: .github/workflows/continuous_integration.yaml
Comment (kylekthompson) [reply .github/workflows/continuous_integration.yaml:18]: > annotated tag

I think this is the nuance I was missing, maybe

Comment (michaelglass) [reply .github/workflows/continuous_integration.yaml:18]: yah `git tag  -a -m "" v1.2.3` or `git tag -s -m "" v1.2.3` (can include a message but 🤷‍♀️)

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:42]: 🙏

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:97]: 🤘

File: .github/workflows/release.yaml
Comment (michaelglass) [inline .github/workflows/release.yaml:188]: 🤘

**Pull request merged**
